tests/test_sparse.py

```diff
@@ -0,0 +1,43 @@
+import errno
+import os
+import subprocess
+import sys
+import tempfile
+from typing import IO
+
+from tests.testsupport import *
+
+
+def test_sparse():
+    file = tempfile.NamedTemporaryFile(delete=False)
+    try:
+        sabctools.sparse(file, 100)
+        assert os.path.getsize(file.name) == 100
+        assert is_sparse(file) is True
+    finally:
+        file.close()
+        os.unlink(file.name)
+
+
+def is_sparse(file: IO) -> bool:
```

Comment:
> Does this also work for sparse files, which are partly filled?

No, it's a specific method I made to test if the whole file is unallocated but I'm pretty sure using st_blocks on Linux/Mac would be sufficient, `return os.stat(file.name).st_blocks*512 < os.path.getsize(file.name)` it's just testing allocated vs size. I think I'll change it when I get a chance.

Using seeks and your example the seeks would reveal the first sector as DATA followed by a HOLE at 4096 (or whatever the sector size is) - hence sparse files can get quite fragmented.

I also wouldn't use the Windows method of calling the fsutil process outside of a test, it was just the simplest method I could find - alternatively I think it would need another C function calling Windows APIs to see if it has the sparse attribute or check for zero regions (FSCTL_QUERY_ALLOCATED_RANGES) it didn't seem worth it since a sparse file is transparent to the application and I'd expect sabs behaviour to be to prefer sparse but it is of no consequence if it is not.

Reply:
I've added a commit which simplifies is_sparse as suggested.

I also merged master so the tests run but I'd expect #85 to be merged first anyway, or this can be merged into it instead.

Reply:
Ouch ... Windows:

```
>>> os.stat(myfile).st_blocks*512 / os.path.getsize(myfile) < 0.9
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'os.stat_result' object has no attribute 'st_blocks'
```

So I tried:

```
>>> myfile = "sparse_or_not.bin"
>>> with open(myfile, "w") as file:
...     file.truncate(11000111)
...
11000111
>>> command = 'fsutil sparse queryflag ' +  myfile
>>> command
'fsutil sparse queryflag sparse_or_not.bin'
>>> import os
>>> os.popen(command).readlines()
['This file is NOT set as sparse\n']
```
... computer says no :-(

Reply:
Yeah that's basically what this method is to solve, on Linux/Mac it already works as we want using truncate and most filesystem will set the size instantly so for those OS's that is all this sparse method is doing.

For Windows there is a little more detail on https://github.com/python/cpython/issues/84091 but basically truncate calls `_chsize_s` which zero fills the file. So for example for every 100MB rar we'd be writing all the NUL/zeros before we even start reassembling the file.

The solution is to use Windows APIs to set the file attributes to make it sparse then set the file length without writing anything (for an empty file, size will be set but 'size on disk' will be 0).